### ISSUE TITLE
Secondary rate limit hit but action didn't retry

### ISSUE THREAD

**Issue by rmunn** [EnricoMi/publish-unit-test-result-action#619]
Last night I had a workflow run where the action succeeded in publishing unit test results, but hit a GitHub secondary rate limit afterwards while accessing the URL `/search/issues?q=type%3Apr+repo%3A%22(ORG)%2F(REPO)%22+(COMMIT_HASH)&per_page=100` (details that could identify the repo removed, sorry). The code that is supposed to wait `secondary_rate_limit_wait_seconds` seconds and try again apparently didn't run: from the logs, it seems that it was looking for a Retry-After header from GitHub but GitHub didn't provide one in this instance. Details (as much as I can provide, sorry it can't be everything) follow:

How the action is called in our workflow:

      - name: Report test results
        uses: EnricoMi/publish-unit-test-result-action@v2.3.0
        if: always()
        with:
          nunit_files: "**/TestResults/TestResults.xml"

Yes, that's an old version. I'd love to update the action to 2.17.0 and try again, but realistically, it's unlikely that I could consistently reproduce hitting the secondary rate limit that we hit last night, so I'm afraid I won't be able to verify whether updating to 2.17.0 would fix the issue (i.e., there might already be a fix for this somewhere between 2.3.0 and 2.17.0, but I won't be able to verify that).

Relevant section of the workflow logs:

```plaintext
2024-08-28 01:15:32 +0000 - publish -  INFO - Available memory to read files: 13.7 GiB
2024-08-28 01:15:32 +0000 - publish -  INFO - Reading NUnit files **/TestResults/TestResults.xml (2 files, 1.1 MiB)
2024-08-28 01:15:32 +0000 - publish -  INFO - Finished reading 2 files in 0.02 seconds
2024-08-28 01:15:32 +0000 - publish -  INFO - Publishing success results for commit (COMMIT_HASH)
2024-08-28 01:15:34 +0000 - publish -  INFO - Created check https://github.com/(ORG)/(REPO)/runs/(RUN_NUMBER)
2024-08-28 01:15:34 +0000 - publish -  INFO - Created job summary
2024-08-28 01:15:34 +0000 - publish - WARNING - Request GET /search/issues?q=type%3Apr+repo%3A%22(ORG)%2F(REPO)%22+(COMMIT_HASH)&per_page=100 failed: Forbidden
2024-08-28 01:15:34 +0000 - publish - WARNING - Request GET /search/issues?q=type%3Apr+repo%3A%22(ORG)%2F(REPO)%22+(COMMIT_HASH)&per_page=100 failed with 403: Forbidden
2024-08-28 01:15:34 +0000 - publish -  INFO - There is no Retry-After given in the response header
2024-08-28 01:15:34 +0000 - publish -  INFO - Response message does not indicate retry-able error
Warning: Request GET /search/issues?q=type%3Apr+repo%3A%22(ORG)%2F(REPO)%22+(COMMIT_HASH)&per_page=100 failed with 403: Forbidden
Traceback (most recent call last):
  File "/action/publish_test_results.py", line 460, in <module>
    main(settings, gha)
  File "/action/publish_test_results.py", line 183, in main
    Publisher(settings, gh, gha).publish(stats, results.case_results, conclusion)
  File "/action/publish/publisher.py", line 197, in publish
    pulls = self.get_pulls(self._settings.commit)
  File "/action/publish/publisher.py", line 209, in get_pulls
    issues = list(self._gh.search_issues(f'type:pr repo:"{self._settings.repo}" {commit}'))
  File "/usr/local/lib/python3.8/site-packages/github/PaginatedList.py", line 56, in __iter__
    newElements = self._grow()
  File "/usr/local/lib/python3.8/site-packages/github/PaginatedList.py", line 67, in _grow
    newElements = self._fetchNextPage()
  File "/usr/local/lib/python3.8/site-packages/github/PaginatedList.py", line 201, in _fetchNextPage
    headers, data = self.__requester.requestJsonAndCheck(
  File "/usr/local/lib/python3.8/site-packages/github/Requester.py", line 354, in requestJsonAndCheck
    *self.requestJson(
  File "/usr/local/lib/python3.8/site-packages/github/Requester.py", line 454, in requestJson
    return self.__requestEncode(cnx, verb, url, parameters, headers, input, encode)
  File "/usr/local/lib/python3.8/site-packages/github/Requester.py", line 528, in __requestEncode
    status, responseHeaders, output = self.__requestRaw(
  File "/action/publish_test_results.py", line 211, in throttled_gh_request_raw
    return gh_request_raw(cnx, verb, url, requestHeaders, input)
  File "/usr/local/lib/python3.8/site-packages/github/Requester.py", line 555, in __requestRaw
    response = cnx.getresponse()
  File "/usr/local/lib/python3.8/site-packages/github/Requester.py", line 127, in getresponse
    r = verb(
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 542, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 529, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 645, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/adapters.py", line 440, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 868, in urlopen
    retries = retries.increment(method, url, response=response, _pool=self)
  File "/action/publish/retry.py", line 101, in increment
    raise GithubException(response.status, content, response.headers)
github.GithubException.GithubException: 403 {"documentation_url": "https://docs.github.com/free-pro-team@latest/rest/overview/rate-limits-for-the-rest-api#about-secondary-rate-limits", "message": "You have exceeded a secondary rate limit. Please wait a few minutes before you try again. If you reach out to GitHub Support for help, please include the request ID 60C0:13D5D6:40861A:78B8D1:66CE7A36."}
```

As you can see, the 403 response from GitHub contained a JSON structure whose `documentation_url` could probably be parsed to figure out that a secondary rate limit was hit, but there was apparently no Retry-After header. Also, this happened *after* the unit test results had already been uploaded; our PR check that verifies unit test results saw the results and gave a green check since all tests passed in that particular run.

Re-running the workflow a few hours later succeeded, so I'm afraid this might be difficult to reproduce. (For example, as I mentioned previously, I could update the action to v2.17.0 and try again, but that probably wouldn't prove anything: we've been running v2.3.0 hundreds of times at this point and this is the first time we've gotten this particular 403 response). Still, I'm reporting this so that you'll at least be able to go look at the relevant v2.3.0 code and see if the "secondary rate limit errors have a Retry-After header" assumption is still present in v2.17.0. Hopefully this is at least a little bit helpful.

**Comment by EnricoMi** [comment]
Thanks for the detailed explanation.

The `/search` end-point has a much higher risk of hitting the secondary rate limit. For that reason, the action moved away from that end-point in [v2.6.0](https://github.com/EnricoMi/publish-unit-test-result-action/releases/tag/v2.6.0).

This particular rate limit error was not recognized by v2.3.0, as you correctly pointed out, but that has been fixed in [v2.9.0](https://github.com/EnricoMi/publish-unit-test-result-action/releases/tag/v2.9.0).

So moving to v2.9.0 and later will fix this issue for you. You could also move to `v2` to use the latest version.

**Comment by rmunn** [comment]
Great; we'll update to the latest version of the action.

Closing issue as it's solved already and there's no work needed to be done (by you, at least) to fix it.
